Shared/ImageLoadingHandler.swift

```diff
@@ -0,0 +1,185 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0
+
+import Foundation
+import Kingfisher
+import UIKit
+
+public enum ImageLoadingError: Error, CustomStringConvertible {
+    case unableToFetchImage
+    case iconUrlNotFound
+    case sizeTooLarge
+
+    public var description: String {
+        switch self {
+        case .unableToFetchImage: return "Favicon Image Error: Unable to fetch image"
+        case .iconUrlNotFound: return "Facicon url not found"
+        case .sizeTooLarge: return "Image download size too large"
+        }
+    }
+}
+
+public struct ImageLoadingConstants {
+    public static let MaximumFaviconSize = 1 * 1024 * 1024 // 1 MiB file size limit
+    public static let ExpirationTime = TimeInterval(60*60*24*7) // Only check for icons once a week
+    public static let NoLimitImageSize = 0 // 0 Means there is no limit for image size
+}
+
+protocol ImageFetcher {
+    func getImageFromCacheOrDownload(with url: URL, limit maxSize: Int,
+                                     completion: @escaping (UIImage?, ImageLoadingError?) -> Void)
+    func saveImageToCache(img: UIImage, key: String)
+    func getImageFromCache(url: URL,
+                           completion: @escaping (UIImage?, ImageLoadingError?) -> Void)
+    func isImageInCache(url: URL) -> Bool
+    func downloadImageOnly(with url: URL,
+                           limit maxSize: Int,
+                           completion: @escaping (UIImage?, Data?, ImageLoadingError?) -> Void)
+    func downloadImageOnly(with url: URL,
```

Comment:
nit: Seems like this 2 functions are very similar except the `maxSize` parameter I was wondering if we could define only one and in the implementation have a default value to avoid repeating the code, something like:
`public func downloadImageOnly(with url: URL,
                                         limit maxSize: Int = 0,
                                         completion: @escaping (UIImage?, Data?, ImageLoadingError?) -> Void) {`